tests/PHPStan/Analyser/nsrt/preg_match_shapes.php

```diff
@@ -306,3 +306,21 @@ function groupsOptional(string $size): void
 	}
 	assertType('array{0: string, 1?: string, 2?: string}', $matches);
 }
+
+function topLevelAlternation(string $size): void
+{
+	if (preg_match('~^(?:(\\d+)x(\\d+)|(\\d+)|x(\\d+))$~', $size, $matches) !== 1) {
+		throw new InvalidArgumentException(sprintf('Invalid size "%s"', $size));
+	}
+	assertType('array{0: string, 1: string, 2: string, 3?: string, 4?: string}', $matches);
+
+	if (preg_match('~^(?:(\\d+)x(\\d+)|(\\d+)|x(\\d+))?$~', $size, $matches) !== 1) {
```

Comment:
Better to divide this in multiple functions, reusing the `$matches` variable like that seems dirty to me.